proto/kuksa/val/v1/types.proto

```diff
@@ -0,0 +1,271 @@
+  // NOTE: This is what you set in order to command the the actuator
+  //       to move.
+  Value target = 30;
+}
+
+////////////////////////
+// Sensor specific
+message Sensor {
+//nothing for now
+}
+
+////////////////////////
+// Attribute specific
+message Attribute {
+  // nothing for now.
+}
+
+
+
+// Value restriction
+//
+// One ValueRestriction{type} for each type, since
+// they don't make sense unless the types match
+//
+message ValueRestriction {
+  oneof restriction_type {
+    ValueRestrictionsString string_restriction = 1;
+    ValueRestrictionInt int_restriciton        = 2; //for all signed VSS integers
+    ValueRestrictionUint uint_restiction       = 3; //for all unsigned VSS integers
+    ValueRestrictionFloat float_restriction    = 4; //for VSS float and double
+  }
+}
+
+
+message ValueRestrictionInt {
+  oneof optional_min {
+    sint64 min = 1;
+  }
+  oneof optional_max {
+    sint64 max = 2;
+  }
+  repeated sint64 allowed_values = 3;
+}
+
+message ValueRestrictionUint {
+  oneof optional_min {
+    uint64 min = 1;
+  }
+  oneof optional_max {
+    uint64 max = 2;
+  }
+  repeated uint64 allowed_values = 3;
+}
+
+message ValueRestrictionFloat {
+  oneof optional_min {
+    double min = 1;
+  }
+  oneof optional_max {
+    double max = 2;
+  }
+  repeated double allowed_values = 3; //allowed for doubles/floats not recommended
+}
+
+message ValueRestrictionDouble {
+  oneof optional_min {
+    double min = 1;
+  }
+  oneof optional_max {
+    double max = 2;
+  }
+  repeated double allowed_values = 3;
+}
+
+
+// min, max doesn't make much sense for a string
+message ValueRestrictionsString {
+  repeated string allowed_values = 3;
+}
+
+
+// VSS Data type of a signal
+//
+// Protobuf doesn't support int8, int16, uint8 or uint16.
+// These are mapped to int32 and uint32 respectively.
+//
+enum DataType {
+  STRING          = 0;
+  BOOLEAN         = 1;
+  INT8            = 2;
+  INT16           = 3;
+  INT32           = 4;
+  INT64           = 5;
+  UINT8           = 6;
+  UINT16          = 7;
+  UINT32          = 8;
+  UINT64          = 9;
+  FLOAT           = 10;
+  DOUBLE          = 11;
+  TIMESTAMP       = 12;
+  STRING_ARRAY    = 20;
+  BOOLEAN_ARRAY   = 21;
+  INT8_ARRAY      = 22;
+  INT16_ARRAY     = 23;
+  INT32_ARRAY     = 24;
+  INT64_ARRAY     = 25;
+  UINT8_ARRAY     = 26;
+  UINT16_ARRAY    = 27;
+  UINT32_ARRAY    = 28;
+  UINT64_ARRAY    = 29;
+  FLOAT_ARRAY     = 30;
+  DOUBLE_ARRAY    = 31;
+}
+
+
+enum FieldType {
+  CURRENT_VALUE = 0;  // this is default, every implementation shall support this
+  TARGET_VALUE  = 1;  // this is a set point of an actuator
+  DATATYPE      = 2;  // VSS datatype of an element as string ("branch" in case of a VSS branch)
+  UNIT          = 3;   // VSS unit of an element as string
+  CHILDREN      = 4;   // Names of all direct children as string
+  ALL_METADATA  = 100; // All available metadata defined in VSS model
+  ALL           = 200; // All available data
+}
```

Comment:
I think it would be better to have a 1:1 mapping between fields and this enum, i.e.

```
message Datapoint {
    Value value = 2;          // Field = VALUE
    DataType data_type = 3;   // Field = DATA_TYPE
    string description = 4;   // Field = DESCRIPTION
    string comment = 10;      // Field = COMMENT
    string deprecation = 11;  // Field = DEPRECATION
    string unit = 12;         // Field = UNIT
    ValueRestriction value_restriction = 13; // Field = VALUE_RESTRICTION
    Actuator actuator   = 30; // Field = ACTUATOR
    Sensor sensor       = 31; // Field = SENSOR
    Attribute attribute = 32; // Field = ATTRIBUTE
}

message Actuator {
    Value target = 1; // Field = ACTUATOR_TARGET
}
```

and

```suggestion
enum Field {
  VALUE = 0;  // this is default, every implementation shall support this
  ACTUATOR_TARGET = 1;  // this is a set point of an actuator
  DATA_TYPE      = 2;  // VSS datatype of an element as string ("branch" in case of a   VSS branch)
  UNIT          = 3;   // VSS unit of an element as string
  DESCRIPTION = 4;
  COMMENT = 5;
  DEPRECATION = 6;
  VALUE_RESTRICTION = 7;
  ACTUATOR = 8;
  SENSOR = 9;
  ATTRIBUTE = 10;

  ALL_METADATA  = 100; // All available metadata defined in VSS model
  ALL           = 200; // All available data

}
```